# Module 4.3: Phylogenetic Trees

```{r setup, include=FALSE}
library(here)
source(here("code","setup.R"))
```

```{r data-import-run, include=FALSE}
# Importing filtered genotypic data
matrix <- read.table(here("data", "FilteredBarley.txt"), sep = "\t", header = TRUE, row.names = 1, check.names = FALSE)
# SNP matrix has to have individuals in rows and markers as columns for the posterior functions
matrix <- t(matrix)

# Importing metadata
metadata <- read_excel(here("data", "BarleyMetadata.xlsx"))
metadata <- metadata[metadata$Individual %in% rownames(matrix),] #Ensuring IDs match
```

```{r data-import, eval=FALSE}
# Importing filtered genotypic data
matrix <- read.table("data/FilteredBarley.txt", sep = "\t", header = TRUE, row.names = 1, check.names = FALSE)
# SNP matrix has to have individuals in rows and markers as columns for the posterior functions
matrix <- t(matrix)

# Importing metadata
metadata <- read_excel("data/BarleyMetadata.xlsx")
```

```{r}
tree <- phyloTree(matrix, treeType = "upgma", distanceType = nei.dist, samples = 1, path = here("output", "tree.txt"))

plot(tree, main = "UPGMA Tree with Bootstrap", cex = 0.1)
```
